File: Problems/Food Delivery System/Java/deliveryAgent/DeliveryAgent.java
package deliveryAgent;

import order.Order;

public class DeliveryAgent {
    private int id;
    private String name;
    private Order order;
    private boolean isAvailable;

    public DeliveryAgent(int id, String name) {
        this.id = id;
        this.name = name;
        this.isAvailable = true; // Default to available
    }

    public int getId() {
        return id;
    }

    public void assignOrder(Order order) {
        this.order = order;
        this.isAvailable = false; // Mark as unavailable when assigned an order
    }

    public void completeOrder() {
        this.order = null;
        this.isAvailable = true; // Mark as available when order is completed
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public String getName() {
        return name;
    }

    public Order getOrder() {
        return order;
    }
}

File: Problems/Food Delivery System/Golang/deliveryagent/delivery_agent.go
package deliveryagent

import "sync"

type DeliveryAgent struct {
	ID          int
	Name        string
	OrderID     string
	isAvailable bool
	mu          sync.RWMutex
}

func NewDeliveryAgent(id int, name string) *DeliveryAgent {
	return &DeliveryAgent{ID: id, Name: name, isAvailable: true}
}

func (d *DeliveryAgent) AssignOrder(orderID string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.OrderID = orderID
	d.isAvailable = false
}

func (d *DeliveryAgent) CompleteOrder() {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.OrderID = ""
	d.isAvailable = true
}

func (d *DeliveryAgent) IsAvailable() bool {
	d.mu.RLock()
	defer d.mu.RUnlock()
	return d.isAvailable
}
